### PR TITLE
CMake: installation of pkgconfig & CMake config file

### PR DESCRIPTION
**Pull request by SpaceIm**
see https://github.com/conan-io/conan-center-index/pull/8930

Usage in CMake for consumers:

```cmake
find_package(cocoyaxi REQUIRED CONFIG)
target_link_libraries(userTarget cocoyaxi::co)
```

I've also added an alias target `cocoyaxi::co` for users relying on `add_subdirectory()`.

And some modifications I had to do because exporting targets require to be clear regarding dependencies:
- rely on CMake targets in order to be able to create a relocatable package (and in the meantime targets are far more robust). I had to bump min CMake version to 3.12 since it requires `CURL::libcurl` target: https://cmake.org/cmake/help/latest/module/FindCURL.html
- I've removed link to zlib since it seems to be a transitive dependency coming from libcurl, not a direct dependency, do you confirm?
- use `find_package(Threads)` and `Threads::Threads` target, it's more robust than raw link to pthread.
- `ws2_32` is not just for MinGW, but for Windows (it comes for free in a Visual Studio environment but it's neater to be explicit).

### CODE REVIEW

**Review by idealvin — APPROVED** [review]
Great job! Thanks!

**Comment by SpaceIm** [comment]
In your top CMakeLists I see:

```cmake
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_SOURCE_DIR}/build/lib)
set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_SOURCE_DIR}/build/lib)
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_SOURCE_DIR}/build/bin)
```

It's really not a good idea.
If you really want to put runtimes in the same folder, just do `set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})`.

**Comment by idealvin** [comment]
@SpaceIm I'm not familar with cmake. Could you help provide a better solution? Thanks!

**Comment by SpaceIm** [comment]
I can change this in this PR or another, as you wish.

**Comment by idealvin** [comment]
> I can change this in this PR or another, as you wish.

ok, just do it in this PR please, thanks!

**Review by idealvin — APPROVED** [review]

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -29,46 +23,48 @@ if(WIN32)
...
-    endif()
+    set_target_properties(co PROPERTIES POSITION_INDEPENDENT_CODE ON)
 endif()
 
+find_package(Threads REQUIRED)
+target_link_libraries(co PUBLIC Threads::Threads)
+
+if(UNIX AND NOT APPLE)
+    target_link_libraries(co PRIVATE dl)
+endif()
```
Comment (SpaceIm) [inline src/CMakeLists.txt:67]: Are you sure that dl is a direct dependency of co?

Comment (idealvin) [reply src/CMakeLists.txt:67]: Yes, co depends on `dlsym()` on unix.

Comment (SpaceIm) [reply src/CMakeLists.txt:67]: Indeed 👍

**Comment by SpaceIm** [comment]
So the scope of the PR was extended a little bit, if it's fine for you I've nothing more to change in this PR.

**Comment by idealvin** [comment]
> So the scope of the PR was extended a little bit, if it's fine for you I've nothing more to change in this PR.

ok, thank you!

**Comment by SpaceIm** [comment]
Maybe update the README to explain how to consume cocoyaxi if using CMake.

**Review by idealvin — APPROVED** [review]

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -100,17 +94,95 @@ else()
...
+install(
+    DIRECTORY ${PROJECT_SOURCE_DIR}/include/co
+    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+)
+
+## pkgconfig file
+if(WIN32)
+    string(APPEND CO_PKG_EXTRA_LIBS " -lws2_32")
+else()
+    if(UNIX AND NOT APPLE)
```
[the remark this answers is not in the document]
Comment (SpaceIm) [reply src/CMakeLists.txt:119]: you don't have to link pthread and dl on macOS

[the remark this answers is not in the document]
Comment (idealvin) [reply src/CMakeLists.txt:119]: ok, thanks!

**Pull request merged**